File: src/CandidatosCore.Core/GestorDeCandidato.cs
namespace CandidatosCore.Core;

using CandidatosCore.Core.Dominio;
using CandidatosCore.Core.Historial;
using CandidatosCore.Core.Notificaciones;

public sealed class GestorDeCandidato
{
    private readonly HistorialCambios historial;
    private readonly NotificadorCambios notificador;

    public GestorDeCandidato(HistorialCambios historial, NotificadorCambios notificador)
    {
        this.historial = historial;
        this.notificador = notificador;
    }

    public void Avanzar(Candidato candidato, string actor, string comentario = "", bool esNotaInterna = false) =>
        Ejecutar(candidato, actor, TipoCambio.Avance, candidato.Avanzar, comentario, esNotaInterna);

    public void Rechazar(Candidato candidato, string actor, string comentario = "", bool esNotaInterna = false) =>
        Ejecutar(candidato, actor, TipoCambio.Rechazo, candidato.Rechazar, comentario, esNotaInterna);

    public void DeshacerUltimo(Candidato candidato, string actor, string comentario = "")
    {
        var registro = historial.DeshacerUltimo(candidato, actor, comentario);
        notificador.Notificar(registro);
    }

    private void Ejecutar(Candidato candidato, string actor, TipoCambio tipo, Action transicion, string comentario, bool esNotaInterna = false)
    {
        var anterior = candidato.ObtenerFaseActual();
        transicion();
        var nuevo = candidato.ObtenerFaseActual();

        var registro = historial.RegistrarCambio(candidato, anterior, nuevo, actor, tipo, comentario, esNotaInterna);
        notificador.Notificar(registro);
    }
}

File: src-ts/CandidatosCore.Core/GestorDeCandidato.ts
import type { Candidato } from "./Dominio/Candidato";
import type { HistorialCambios } from "./Historial/HistorialCambios";
import { TipoCambio } from "./Historial/TipoCambio";
import type { NotificadorCambios } from "./Notificaciones/NotificadorCambios";

export class GestorDeCandidato {
  private readonly historial: HistorialCambios;
  private readonly notificador: NotificadorCambios;

  constructor(historial: HistorialCambios, notificador: NotificadorCambios) {
    this.historial = historial;
    this.notificador = notificador;
  }

  public avanzar(
    candidato: Candidato,
    actor: string,
    comentario: string
  ): void {
    this.ejecutar(
      candidato,
      actor,
      TipoCambio.Avance,
      () => candidato.avanzar(),
      comentario
    );
  }

  public rechazar(
    candidato: Candidato,
    actor: string,
    comentario: string
  ): void {
    this.ejecutar(
      candidato,
      actor,
      TipoCambio.Rechazo,
      () => candidato.rechazar(),
      comentario
    );
  }

  public deshacerUltimo(
    candidato: Candidato,
    actor: string,
    comentario: string
  ): void {
    const registro = this.historial.deshacerUltimo(
      candidato,
      actor,
      comentario
    );
    this.notificador.notificar(registro);
  }

  private ejecutar(
    candidato: Candidato,
    actor: string,
    tipo: TipoCambio,
    transicion: () => void,
    comentario: string
  ): void {
    const anterior = candidato.obtenerFaseActual();
    transicion();
    const nuevo = candidato.obtenerFaseActual();

    const registro = this.historial.registrarCambio(
      candidato,
      anterior,
      nuevo,
      actor,
      tipo,
      comentario
    );
    this.notificador.notificar(registro);
  }
}
